aten/src/ATen/native/BackendSelectRegister.cpp

```diff
@@ -0,0 +1,631 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/TensorOptions.h>
+
```

Comment:
It would be good to have a comment here explaining the function of this file. I also wonder a bit about whether or not we should do things to reduce duplication here. For example, the types for each of these functions has to be written twice (the function definition, and then the template argument) and it would be great if we only had to write it once. 

@smessmer, if I have a template function `template <T> f(T&&)` and I call it from some other function `g(S x) { 
f(x); }`, is the `T` in `f` guaranteed to deduce to `S` no matter what S is?

Reply:
Basic rule: `&` and `&&` together decay to `&`.

Given a non-reference concrete type `A`.
- If `S` is a lvalue reference (i.e. `S=A&`), then `T=S=A&`.
- If `S` is a non-reference (i.e. `S=A`), then `T=S&=A&` because a non-reference `A` can be passed to functions taking `A` or `A&` but not to functions taking `A&&`. `T=S` would lead to `T&& = A&&` and the argument type wouldn't match the call site. So instead, `T=S&=A&` and `T&& = (A&)&& = A&` decays, the argument can be passed in.
- If `S` is an rvalue reference (i.e. `S=A&&`), then `T=decay_t<S>=A`. This allows `T&& = A&&` and allows moving in the argument.

Note that this assumes that your call to `f(x)` forwards, i.e. looks like `f(std::forward<S>(x))`. In the way you've written it, you're always passing `x` by value, which means `T` will always be `S&`.

Reply:
For simplifying this code here, the easiest way is probably to use a template and specify the type arguments explicitly instead of inferring them.

Reply:
Hmm, but I don't think we ever have `A&&` arguments today, so maybe it would work out?

Reply:
why do callUnboxedWithDispatchKey templates not match the decltype of the function?  Is it because we want to apply some additional checking?

Reply:
> why do callUnboxedWithDispatchKey templates not match the decltype of the function? Is it because we want to apply some additional checking?

It does match. It just looks a bit different due to how callUnboxedWithDispatchKey is defined:

  ```
template<class Return, class... Args>

 Return callUnboxedWithDispatchKey(const OperatorHandle& op, DispatchKey dispatchKey, Args... args) const;

```

So when we have a call like an `eye`
`op.callUnboxedWithDispatchKey<Tensor, int64_t, const TensorOptions &>(key, n, options);`
`Tensor` - `Return`
`int64_t, const TensorOptions & `- `Args`
`key` - `DispatchKey dispatchKey`
`n, options`- `Args... args`

Reply:
> It would be good to have a comment here explaining the function of this file. I also wonder a bit about whether or not we should do things to reduce duplication here. For example, the types for each of these functions has to be written twice (the function definition, and then the template argument) and it would be great if we only had to write it once.
> 
> @smessmer, if I have a template function `template <T> f(T&&)` and I call it from some other function `g(S x) { f(x); }`, is the `T` in `f` guaranteed to deduce to `S` no matter what S is?

Added a comment